tests/spell_check3.k

```diff
@@ -0,0 +1,318 @@
+    role SpellChecker {
+        public void advanceToNextWord() {
+            if(isFinished()) return;
+
+            // Advance past current word
+            while(isAlphabetic(currentCharacter())) {
+                setPosition(currentPosition() + 1);
+                if(isFinished()) return;
+            }
+
+            // Advance past non-alphabetic characters
+            while(!isAlphabetic(currentCharacter())) {
+                setPosition(currentPosition() + 1);
+                if(isFinished()) return;
+            }
+
+            Dictionary.checkCurrentWord();
+        }
+
+        public String currentWord() {
+            if(isFinished()) return "";
+
+            String output = "";
+            int pos = currentPosition().clone;
+
+            while(isAlphabetic(Text.characterAt(pos))) {
+                output = output + Text.characterAt(pos++);
+                // NOTE: Cannot call Text.getText().length() here.
+                if(pos >= Text.length()) return output;
+            }
+
+            return output;
+        }
+
+        private String currentCharacter() {
+            return Text.characterAt(currentPosition());
+        }
+
+        private boolean isFinished() {
+            return currentPosition() >= Text.length();
+        }
+
+        private boolean isAlphabetic(String character) {
+            assert(character.length() == 1, "character string length not equal to one.");
+            return character == "-" || 
+                   (character >= "a" && character <= "z") ||
+                   (character >= "A" && character <= "Z");
+        }
+    } requires {
+        int currentPosition() const;
+        void setPosition(int pos);
+    }
+
+    /////////////////////////////////////////
+    role Text {
+        public String characterAt(int pos) {
+            return getText().substring(pos, pos+1);
+        }
+        public int length() {
+            return getText().length();
+        }
+    } requires {
+        String getText() const;
+        void splice(int startIndex, int endIndex, String text);
+    }
+
+    /////////////////////////////////////////
+    stageprop Dictionary {
+        public void checkCurrentWord() {
+            String word = SpellChecker.currentWord();
+            List<String> suggestions = suggestSpelling(word);
+
+            if(suggestions.size() > 0) {
+                int pos = SpellChecker.currentPosition();
+
+                // TODO: User input and selection
+                Text.splice(pos, pos + word.length(), "[" + suggestions.get(0) + "]");
+            }
+
+            SpellChecker.advanceToNextWord();
+        }
+    } requires {
+        List<String> suggestSpelling(String word) const;
+    }
+
+    /////////////////////////////////////////
+
+    public SpellCheck(Text text) {
+        Text = text;
+        SpellChecker = this;
+        Dictionary = new SimpleDictionary();
+    }
+
+    public void start() {
+        setPosition(0);
```

Comment:
> Den 18/01/2016 kl. 17.57 skrev Matt Browne <notifications@github.com>:
> 
> I am a bit concerned that the current restriction in trygve from accessing Context members from role methods may cause programmers to nearly always make the context play a role, merely to make context properties accessible - whether that matches the mental model or not.

In a real use case there are no “context properties.” If you find you are always making the Context a Role then you are likely missing at least one Role in the first place. I think that the next step that most programmers will take is to create a “leftovers Role” for all the class-oriented stuff that didn’t make it into the use case model. This new restriction will help you think more clearly in terms of Roles and focus more on the action than returning to the recurring discussions at the structural level about double. Referring to Context properties has the danger of making the use case stateful (since Contexts are not precluded from having state) and, more to the point, about their methods creating invisible links in the interaction graph between Roles. It’s simply not DCI any more.

It’s long been on my docket to make this change, but what tipped the balance was that Trygve noticed it while playing around with the environment and wrote directly to me to explicitly express his concern that Role methods could access the Context. So I made it a priority and changed it.

Hopefully Trygve can chime in here with a pithy insight from another angle